apps/services/auth-public-api/src/app/modules/delegations/meDelegations.controller.ts

```diff
@@ -0,0 +1,313 @@
+  @Put(':delegationId')
+  @ApiOkResponse({ type: DelegationDTO })
+  @ApiBadRequestResponse({ type: HttpProblemResponse })
+  @ApiNotFoundResponse({ type: HttpProblemResponse })
+  @ApiUnauthorizedResponse({ type: HttpProblemResponse })
+  @ApiForbiddenResponse({ type: HttpProblemResponse })
+  @ApiInternalServerErrorResponse()
+  async update(
+    @CurrentUser() user: User,
+    @Body() delegation: UpdateDelegationDTO,
+    @Param('delegationId') delegationId: string,
+  ): Promise<DelegationDTO | null> {
+    if (!(await this.validateScopesAccess(user.scope, delegation.scopes))) {
+      throw new BadRequestException(
+        'User does not have access to the requested scopes.',
+      )
+    }
+
+    if (!this.validateScopesPeriod(delegation.scopes)) {
+      throw new BadRequestException(
+        'If scope validTo property is provided it must be in the future',
+      )
+    }
+
+    return this.auditService.auditPromise<DelegationDTO | null>(
+      {
+        user,
+        namespace,
+        action: 'update',
+        resources: (delegation) => delegation?.id ?? '',
+        meta: { fields: Object.keys(delegation) },
+      },
+      this.delegationsService.update(user.nationalId, delegation, delegationId),
+    )
+  }
+
+  @Scopes(AuthScope.writeDelegations)
+  @FeatureFlag(Features.customDelegations)
+  @Delete(':delegationId')
+  @HttpCode(204)
+  @ApiNoContentResponse()
+  @ApiUnauthorizedResponse({ type: HttpProblemResponse })
+  @ApiForbiddenResponse({ type: HttpProblemResponse })
+  @ApiInternalServerErrorResponse()
+  async delete(
+    @CurrentUser() user: User,
+    @Param('delegationId') delegationId: string,
+  ): Promise<void> {
+    await this.auditService.auditPromise(
+      {
+        user,
+        namespace,
+        action: 'deleteFrom',
+        resources: delegationId,
+      },
+      this.delegationsService.delete(user.nationalId, delegationId),
+    )
+  }
+
+  /**
+   * Validates that the delegation scopes belong to user and are valid for delegation
+   * @param userScopes user scopes from the currently authenticated user
+   * @param requestedScopes requested scopes from a delegation
+   * @returns
+   */
+  private async validateScopesAccess(
+    userScopes: string[],
+    requestedScopes: UpdateDelegationScopeDTO[],
+  ): Promise<boolean> {
+    if (!requestedScopes || requestedScopes.length === 0) {
+      return true
+    }
+
+    for (const scope of requestedScopes) {
+      // Delegation scopes need to be associated with the user scopes
+      if (!userScopes.includes(scope.name)) {
+        return false
+      }
+    }
+
+    // Check if the requested scopes are valid
+    const scopes = requestedScopes.map((scope) => scope.name)
+    const allowedIdentityResources = await this.resourcesService.findAllowedDelegationIdentityResourceListForUser(
+      scopes,
+    )
+    const allowedApiScopes = await this.resourcesService.findAllowedDelegationApiScopeListForUser(
+      scopes,
+    )
```

Comment:
Wondering if the count should be done on the database layer to avoid loading bunch of objects into memory.


But this probably won't be a lot of data, so it might pass.

Reply:
Good comment. I created a count function to return just the count from db. I also just removed the `identityResource` part as we are not using it.